posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,625 @@
+**It is not a string**. It shouldn't have been called \*String. It is a sequence of bytes,
+carries no encoding information and is **pinned memory** (cannot be moved by the GC). As such, it doesn't require
+copying when dealing with the FFI. It is quite efficient and has a large API, but (obviously) lacks text processing
+facilities, because it has no knowledge of unicode. Most operations work on `Word8` boundaries.
+
+The definition for strict ByteString is (as of [0.12.1.0](https://hackage.haskell.org/package/bytestring-0.12.1.0)):
+
+```hs
+data ByteString = BS {-# UNPACK #-} !(ForeignPtr Word8) -- payload
+                     {-# UNPACK #-} !Int                -- length
+```
+
+This allows, similar to Text, slicing without copying memory (through pointer arithmetic and the length field).
+
+And the lazy counterpart, which looks similar to lazy Text:
+
+```hs
+data ByteString = Empty
+                | Chunk  {-# UNPACK #-} !S.StrictByteString ByteString
+```
+
+There is an API variant [Data.ByteString.Char8](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Char8.html),
+which allows operations to work on `Char` boundaries. However, it can be misleading to newcomers, because it actually truncates
+all Chars to 8 bits. **You should avoid this, unless you know what you are doing.** It is more likely that you are looking
+for decoding libraries, where you can specify which encoding to use, e.g. [bytestring-encoding](https://hackage.haskell.org/package/bytestring-encoding-0.1.2.0/docs/Data-ByteString-Encoding.html#v:decode).
+
+It also has to be noted that pinned memory can cause memory fragmentation for lots of small ByteStrings
+(this is also discussed in [Fixing 'FilePath' in Haskell](https://hasufell.github.io/posts/2022-06-29-fixing-haskell-filepaths.html)).
+An alternative type is `ShortByteString`, which will be discussed next.
+
+#### ByteString summary
+
+Invariants:
+
+- pinned memory
+
+Useful for:
+
+- large data
+- dealing with raw bytes (e.g. web servers)
+- dealing with C FFI
+- storing non-unicode encodings e.g. via a newtype wrapper
+- fast parsers, see the excellent blog post from Chris Done on [Fast Haskell: Competing with C at parsing XML](https://chrisdone.com/posts/fast-haskell-c-parsing-xml/)
+
+Not so useful for:
+
+- dealing with unicode or human readable text
+- dealing with lots of small byte sequences
+
+Lazy variants, again, are useful for streaming and incremental processing, as the strict variant requires the whole content to be in memory.
+
+### ShortByteString
+
+This type is from the bytestring package as well and lives under
+[Data.ByteString.Short](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Short.html).
+
+It has the same API as `ByteString` since [0.11.3.0](https://hackage.haskell.org/package/bytestring-0.11.3.0/changelog),
+so can be used as a drop-in replacement. The main difference is that it is *unpinned memory*, so causes no
```

Comment:
This is not strictly correct: `ByteArray` could be allocated as pinned. 
```suggestion
so can be used as a drop-in replacement. The main difference is that it could be backed by *unpinned memory*, so causes no
```

Reply:
However, using the default API to construct ShortByteString's, we always end up with unpinned ones, no?

Reply:
Yes